spatial/kdtree/kdtree.go

```diff
@@ -0,0 +1,465 @@
+	if n == nil {
+		return &Node{
+			Point:    c,
+			Plane:    d,
+			Bounding: nil,
+		}
+	}
+
+	d = (n.Plane + 1) % Dim(c.Dims())
+	if c.Compare(n.Point, n.Plane) <= 0 {
+		n.Left = n.Left.insert(c, d)
+	} else {
+		n.Right = n.Right.insert(c, d)
+	}
+
+	return n
+}
+
+func (n *Node) insertBounded(c Extender, d Dim, bounding bool) *Node {
+	if n == nil {
+		var b *Bounding
+		if bounding {
+			b = c.Extend(b)
+		}
+		return &Node{
+			Point:    c,
+			Plane:    d,
+			Bounding: b,
+		}
+	}
+
+	if bounding {
+		n.Bounding = c.Extend(n.Bounding)
+	}
+	d = (n.Plane + 1) % Dim(c.Dims())
+	if c.Compare(n.Point, n.Plane) <= 0 {
+		n.Left = n.Left.insertBounded(c, d, bounding)
+	} else {
+		n.Right = n.Right.insertBounded(c, d, bounding)
+	}
+
+	return n
+}
+
+// Len returns the number of elements in the tree.
+func (t *Tree) Len() int { return t.Count }
+
+// Contains returns whether a Comparable is in the bounds of the tree. If no bounding has
+// been constructed Contains returns true.
+func (t *Tree) Contains(c Comparable) bool {
+	if t.Root.Bounding == nil {
+		return true
+	}
+	return t.Root.Contains(c)
+}
+
+var inf = math.Inf(1)
+
+// Nearest returns the nearest value to the query and the distance between them.
+func (t *Tree) Nearest(q Comparable) (Comparable, float64) {
+	if t.Root == nil {
+		return nil, inf
+	}
+	n, dist := t.Root.search(q, inf)
+	if n == nil {
+		return nil, inf
+	}
+	return n.Point, dist
+}
+
+func (n *Node) search(q Comparable, dist float64) (*Node, float64) {
+	if n == nil {
+		return nil, inf
+	}
+
+	c := q.Compare(n.Point, n.Plane)
+	dist = math.Min(dist, q.Distance(n.Point))
+
+	bn := n
+	if c <= 0 {
+		ln, ld := n.Left.search(q, dist)
+		if ld < dist {
+			dist = ld
+			bn = ln
+		}
+		if c*c < dist {
+			rn, rd := n.Right.search(q, dist)
+			if rd < dist {
+				bn, dist = rn, rd
+			}
+		}
+		return bn, dist
+	}
+	rn, rd := n.Right.search(q, dist)
+	if rd < dist {
+		dist = rd
+		bn = rn
+	}
+	if c*c < dist {
+		ln, ld := n.Left.search(q, dist)
+		if ld < dist {
+			bn, dist = ln, ld
+		}
+	}
+	return bn, dist
+}
+
+// ComparableDist holds a Comparable and a distance to a specific query. A nil Comparable
+// is used to mark the end of the heap, so clients should not store nil values except for
+// this purpose.
+type ComparableDist struct {
+	Comparable Comparable
+	Dist       float64
+}
+
+// Heap is a max heap sorted on Dist.
+type Heap []ComparableDist
+
+func (h *Heap) Max() ComparableDist  { return (*h)[0] }
+func (h *Heap) Len() int             { return len(*h) }
+func (h *Heap) Less(i, j int) bool   { return (*h)[i].Comparable == nil || (*h)[i].Dist > (*h)[j].Dist }
+func (h *Heap) Swap(i, j int)        { (*h)[i], (*h)[j] = (*h)[j], (*h)[i] }
+func (h *Heap) Push(x interface{})   { (*h) = append(*h, x.(ComparableDist)) }
+func (h *Heap) Pop() (i interface{}) { i, *h = (*h)[len(*h)-1], (*h)[:len(*h)-1]; return i }
+
+// NKeeper is a Keeper that retains the n best ComparableDists that it is called to Keep.
```

Comment:
There's something wrong with `... that it is called to Keep`. Or is it a play on the capital `K`?

Reply:
'called' in this sense is 'demand', however, it is grammatically slightly off; it should be 'that has been called to Keep'. It is a small play on word. I can change it to more normal phrasing:
```
// NKeeper is a Keeper that retains the n best ComparableDists that have been passed to Keep.
```

Reply:
No, I think it's fine to keep as it is.

Reply:
I changed it with the rest.